### ISSUE TITLE
Different color in the calendar for days affected by the booking restriction

### ISSUE THREAD

**Issue by PeteJackson57** [meeting-room-booking-system/mrbs-code#3610]
Hey there,

we are running MRBS 1. 11.4. Our booking options are quite limited, we only want someone to occupy a room for a maximum of 8 weeks in the future. Of course we have set this for the area. Unfortunately, our users keep telling us that they don't see it directly in the calendar. I was wondering whether it would be possible to color the period differently in the calendar, as is the case with holidays and weekends, for example. 

I'm sorry if there's already a threat about this, but I couldn't find anything. 

Thank you!

**Comment by campbell-m** [comment]
No, there isn't already a thread about this as far as I'm aware.

I assume you mean you don't allow users to book more than 8 weeks ahead rather than not allowing them to have more than 8 weeks' worth of bookings outstanding.  It seems a sensible request and I've changed the label from "support request" to "enhancement".

It's not totally trivial though as I think it would probably also need some JavaScript/Ajax to take account of the fact that time is advancing all the time.  Say for example the policy was that you couldn't book more than 1 hour in advance.  If I brought up the calendar for a room at say 14:59 then the 16:00 slot would be styled differently.  But by the time it gets to 15:01 then I could now book it.

**Comment by PeteJackson57** [comment]
Hello Campbell,
thanks for your answer. We only book our rooms on a daily basis.  I solved it this way first and added the days to the Holiday array according to our booking rules. I know it's not a pretty solution but it works. Could you tell me where it checks if the day is a holiday or not takes place? I'm thinking about writing these days into a special array so i can create an own styling class for this days.
`
$currentDate = new DateTime();
$startDate = new DateTime(date('Y-01-01'));
$endDate = new DateTime(date('Y-12-31'));

while ($startDate <= $endDate) {

    //Check that the date is at least 7 days in the future and no more than 8 weeks (56 days)
   
	$difference = $currentDate->diff($startDate);
    $daysDifference = $difference->format('%r%a');
    
    if ($daysDifference >= 7 && $daysDifference <= 56) {
    } else {
        // Add Year to the Array
        $year = $startDate->format('Y');
        if (!isset($holidays[$year])) {
            $holidays[$year] = array();
        }
        $holidays[$year][] = $startDate->format('Y-m-d');
    }
    
    $startDate->modify('+1 day');
}`

**Comment by campbell-m** [comment]
The easiest thing to do is modify the function `get_date_classes()` which is at the begining of functions_table.inc.  Then just add something along the lines of

~~~php
  if ($date->isOutOfBounds())
  {
    $result[] = 'out_of_bounds';   // your new class
  }
~~~

Then add an `isOutOfBounds()` method to lib/MRBS/DateTime.php.  If you want to avoid hard-coding 7 days and 8 weeks into your code you could pick up the global variables `$min_create_ahead_enabled`, `$max_create_ahead_enabled`, `$min_create_ahead_secs` and `$min_create_ahead_secs`.

Then finally modify the block in functions_table.inc around line 190 that looks like

~~~php
      // Don't provide a link if the slot doesn't really exist or if the user is logged in, but not a booking admin,
      // and it's a holiday/weekend and bookings on holidays/weekends are not allowed.  (We provide a link if they
      // are not logged in because they might want to click and login as an admin).
      if ($is_invalid ||
          ((null !== session()->getCurrentUser()) && !is_book_admin($query_vars['new_times']['room']) &&
            (($prevent_booking_on_holidays && in_array('holiday', $classes)) ||
             ($prevent_booking_on_weekends && in_array('weekend', $classes)))))
      {
        $html .= '<span class="not_allowed"></span>';
      }
~~~

to also check whether 'out_of_bounds' is one of `$classes`.
